File: Chapter 1/2. Minesweeper/p10189.py
"""
Online Judge: Problem 10189 - Minesweeper

Judge Version: PYTH3 3.5.1 - Python 3
Best Run Time: 0.010 Seconds

"""

import sys

# For local testing with a text file including sample input
# Uncomment to use local testing
"""
if sys.stdin.isatty():
    sys.stdin = open("p10189.txt", "r")
"""

# All 8 possible neighboring positions around a cell:
# top-left, top, top-right, left, right, bottom-left, bottom, bottom-right
dirs = ((-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1))

def main():
    """
    Main execution loop.
    """

    # Read all input lines at one time and store in a list
    lines = [line.rstrip() for line in sys.stdin.read().splitlines() if line.strip() != ""]

    # Variable used to store entire final resulting output 
    final_output= []
    
    # Index used to keep track of loop through saved input
    idx = 0

    # Keep track of the field number being processed
    # Required for output
    field_cnt = 0

    # Process fields until input  dimensions "0 0" are provided
    while idx < len(lines):

        # Read the dimensions of the field
        n, m = map(int, lines[idx].split())
        idx += 1

        # Field dimensions "0 0" indicate the end of input
        # Exit the loop if this is the case
        if n == 0 and m == 0:
            break

        # Store the field from the input
        # Read the field row by row and store as a list
        field = [list(lines[idx + r]) for r in range(n)]
        idx += n

        # Create a grid with the same dimensions initialized with zeros
        # Used to store counts of mines adjacent to the cells
        counter = [[0]*m for _ in range(n)]

        # Traverse every cell in the field
        for row_num in range(n):
            row = field[row_num]
            for col_num, cell in enumerate(row):

                # If the current cell contains a mine,
                # update all valid neighboring cells
                if cell == '*':

                    # Check all 8 neighboring cells of current cell
                    for dir_row, dir_col in dirs:
                        n_row, n_col = row_num + dir_row, col_num + dir_col

                        # Ensure the neighbor cell being checked is within the field dimemsions and is not a mine
                        # Add 1 to the current value stored in the number grid
                        if 0 <= n_row < n and 0 <= n_col < m and field[n_row][n_col] != '*':
                            counter[n_row][n_col] += 1

        # Increment the field number counter
        field_cnt += 1

        # Print a blank line between consecutive fields after the first one
        if field_cnt > 1:
            final_output.append("") 

        # Add the current field number to the final output list
        final_output.append(f"Field #{field_cnt}:")

        # Build current field and add to final output list
        # Use both field and counter lists
        # Mines remain '*'
        # Empty cells become their adjacent mine count instead of '.'
        for row in range(n):
            final_output.append(''.join('*' if field[row][col] == '*' else str(counter[row][col]) for col in range(m)))

    print("\n".join(final_output))

if __name__ == "__main__":
    main()

File: Chapter 1/2. Minesweeper/p10189.java
/*
* Online Judge: Problem 10189 - Minesweeper
*
* Judge Version: JAVA 1.8.0 - OpenJDK Java
* Best Run Time: 0.040 Seconds
*
*/

import java.io.BufferedReader;
import java.io.InputStreamReader;
//import java.io.FileInputStream;
//import java.io.File;

public class Main {

    // All 8 possible neighboring positions around a cell:
    // top-left, top, top-right, left, right, bottom-left, bottom, bottom-right
    static final int[] dir_x = {-1,-1,-1,0,0,1,1,1};
    static final int[] dir_y = {-1,0,1,-1,1,-1,0,1};

    public static void main(String[] args) throws Exception {

        // For local testing with a text file including sample input
        // Uncomment to use local testing
        /*File testFile = new File("p10189.txt");
        if (testFile.exists()) {
            System.setIn(new FileInputStream(testFile));
        } */
        

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder out = new StringBuilder();

        
        // Keep track of the field number being processed
        // Required for output
        int field_cnt = 0;

        // Variables used for field dimensions
        int n, m;

        // Variables used for neighboring cells
        int n_row, n_col;
        
        // Variable for input line
        String line;

        // Process fields until input dimensions "0 0" are provided
        while (true) {
            line = br.readLine().trim();

            // Read the dimensions of the field
            String[] parts = line.split("\\s+");
            n = Integer.parseInt(parts[0]);
            m = Integer.parseInt(parts[1]);

            // Field dimensions "0 0" indicate the end of input
            // Exit the loop if this is the case
            if (n == 0 && m == 0) {
                break;
            }

            // Initialize field array
            char[][] field = new char[n][m];

            // Create a grid with the same dimensions
            // Used to store counts of mines adjacent to the cells
            // Already initialized with zeros
            int[][] counter = new int[n][m];

           
            // Store the field from the input
            for (int i = 0; i < n; i++) {
                
                char[] row = br.readLine().toCharArray();
               
                for (int j = 0; j < m; j++) {
                    field[i][j] = row[j];
                }
            }


            // Traverse every cell in the field
            for (int row_num = 0; row_num < n; row_num++) {

                for (int col_num = 0; col_num < m; col_num++) {

                    // If the current cell contains a mine,
                    // update all valid neighboring cells
                    if (field[row_num][col_num] == '*') {

                        // Check all 8 neighboring cells of current cell
                        for (int d = 0; d < 8; d++) {
                            n_row = row_num + dir_x[d];
                            n_col = col_num + dir_y[d];

                            // Ensure the neighbor cell being checked is within the field dimemsions and is not a mine
                            // Add 1 to the current value stored in the number grid
                            if (0 <= n_row && n_row < n && 0 <= n_col && n_col < m && field[n_row][n_col] != '*') {
                                counter[n_row][n_col]++;
                            }
                        }
                    }                
                }
            }

            // Print a blank line between consecutive fields after the first one
            if (field_cnt > 0) {
                out.append('\n');
            }   

            // Increment the field count variable
            // Add the current field number to the final output
            out.append("Field #").append(++field_cnt).append(":\n");


            // Build final field
            // Use both field and counter arrays
            // Mines remain '*'
            // Empty cells become their adjacent mine count instead of '.'
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (field[i][j] == '*') {
                        out.append('*');
                    } else {
                        out.append(counter[i][j]);
                    }
                }
                out.append('\n');
            }
        }
        System.out.print(out.toString());
    }
}